File: kubwa-annotations/src/main/java/com/wesleyelliott/kubwa/annotation/CreditCard.java
package com.wesleyelliott.kubwa.annotation;

import com.wesleyelliott.kubwa.rule.CreditCardRule;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.Target;

import static java.lang.annotation.ElementType.ANNOTATION_TYPE;
import static java.lang.annotation.ElementType.CONSTRUCTOR;
import static java.lang.annotation.ElementType.FIELD;
import static java.lang.annotation.ElementType.METHOD;
import static java.lang.annotation.ElementType.PARAMETER;
import static java.lang.annotation.ElementType.TYPE;
import static java.lang.annotation.RetentionPolicy.RUNTIME;

/**
 * Created by wesley on 2016/07/28.
 */
@ValidateUsing(CreditCardRule.class)
@Retention(RUNTIME)
@Target({ METHOD, FIELD, ANNOTATION_TYPE, CONSTRUCTOR, PARAMETER , TYPE})
public @interface CreditCard {
    int errorMessage();
    String name() default "creditCardError";
    CreditCardRule.Type[] types() default {CreditCardRule.Type.VISA, CreditCardRule.Type.MASTERCARD};

    @Target(TYPE)
    @Retention(RUNTIME)
    @Documented
    public @interface List {
        CreditCard[] value();
    }

}

File: kubwa-annotations/src/main/java/com/wesleyelliott/kubwa/annotation/CreditCard.kt
package com.wesleyelliott.kubwa.annotation

import com.wesleyelliott.kubwa.rule.CreditCardRule

/**
 * Created by wesley on 2016/07/28.
 */
@ValidateUsing(CreditCardRule::class)
@Retention(AnnotationRetention.RUNTIME)
@Target(
    AnnotationTarget.FUNCTION,
    AnnotationTarget.PROPERTY_GETTER,
    AnnotationTarget.PROPERTY_SETTER,
    AnnotationTarget.FIELD,
    AnnotationTarget.ANNOTATION_CLASS,
    AnnotationTarget.CONSTRUCTOR,
    AnnotationTarget.VALUE_PARAMETER,
    AnnotationTarget.CLASS,
    AnnotationTarget.FILE
)
annotation class CreditCard(
    val errorMessage: Int,
    val name: String = "creditCardError",
    val types: Array<CreditCardRule.Type> = [CreditCardRule.Type.VISA, CreditCardRule.Type.MASTERCARD]
) {

    @Target(AnnotationTarget.CLASS, AnnotationTarget.FILE)
    @Retention(AnnotationRetention.RUNTIME)
    @MustBeDocumented
    annotation class List(vararg val value: CreditCard)

}
